### ISSUE TITLE
Eliminate the global variable objective_function_penalty_base

### ISSUE THREAD

**Issue by tholden** [DynareTeam/dynare#1051]
This causes problems when the estimation function is evaluated in parallel, using fmincon's 'UseParallel','always' or similar. Further problems with parallel objectives are given in this issue: https://github.com/DynareTeam/dseries/issues/10

**Comment by stepan-a** [comment]
Yes this is the last global in ```dsge_likelihood```. We have at least to precise in the manual that this code cannot be called in parallel. I think it would be a lot of work to pass this variable as an input argument. Maybe we could put the global in a if statement, but I don't know what would be the cost of this solution. Another possibility is to write a version of ```dsge_likelihood```without endogenous penalty (which is only used with ```mode_compute=4,5```.

**Comment by JohannesPfeifer** [comment]
Is there any reason we do not put it as a field into one of the structures we already use? For example into options?

**Comment by tholden** [comment]
dsge_likelihood ought to support being called in parallel really. I've been using CMAES on a multi-modal objective function, with the main loop in parallel. Without the parallelisation it would really be prohibitively slow. As a work-around, I have:
```
if isempty( objective_function_penalty_base ) || objective_function_penalty_base <= 1
	objective_function_penalty_base = Inf;
end
if isstruct( DynareDataset )
    dseries( 'initialize' );
    DynareDataset = dseries( DynareDataset );
end
```
There's also a required change to dseries to get this to work, to support initializing from a struct.

**Comment by JohannesPfeifer** [comment]
I fully agree, but this will be tricky. I am willing to do the change, but we need to find out how to do this. The penalty is sometimes updated. By being global this update is available in later calls. When having the penalty as an input, one needs to make sure that any update is passed to later calls. But this path dependeny is fundamentally incompatible with a parallel context. Maybe we should think about phasing this change in penalties out. @stepan-a @MichelJuillard what exactly is the purpose of this penalty?

**Comment by stepan-a** [comment]
We cannot remove the penalty for ```mode_compute=4,5```. I think that it should be possible to pass it as an argument but it's not simple because we do not systematically update its value (for instance we do not update the penalty when calling numgrad routines (so we would have to pass also a flag to indicate if the penalty bas has to be updated).

The endogenous penalty helps the gradient based methods to find a smooth escape from problematic regions (where BK conditions are not satisfied, a real steady state does not exist, ...).

**Comment by JohannesPfeifer** [comment]
If I understand it correctly, the base value of ```objective_function_penalty_base``` is hard-coded in ```dynare_estimation_init``` to be 1e8 and reset to Inf in ```random_walk_metropolis_hastings```. In between those two statements, ```mode_compute=4,5``` adjust the penalty endogenously. In ```dsge_likelihood.m``` itself, ```objective_function_penalty_base``` is read-only. That means, we should be easily able to transform this global variable into an input that is passed to ```dsge_likelihood.m```. @stepan-a If you agree, I would try to do it this way.

**Comment by stepan-a** [comment]
Your understanding is correct. But I would wait for @MichelJuillard opinion before starting to code... He may remember why he put back this variable as a global.

**Comment by MichelJuillard** [comment]
As stated by @stepan-a  ```objective_function_penalty_base``` helps dealing with the constraints (solving steady state, unicity of RE solution, parameters inside their boundaries, likelihood being real, etc ....) with gradient based methods.
Using a constant penalty creates walls or cliffs in the objective function that give a hard time to the gradient class of optimizers.  Optimizers for constrained optimization and other types will deal with the constraints differently and don't need the penalty.
I think now that it was a design error from my part to put the computation of the smooth penalty in the general objective function ```dsge_likelihood``` when the need occurs only for some optimizers. The hack with the global variable came from my reluctance to change Sims code in ```csminwel```.
We have also to try to preserve backward compatibility for the case where the user supplies his own routine by indicating a function name (or handle) in option ```mode_compute```
As a first step to solve the problem raised by @tholden, I suggest
1. remove ```global objective_function_penaly_base```
2. for each exception in ```dsge_likelihood```, return the variable part of the penalty in ```info(2)``` that was not used by ```dsge_likelihood``` until now
3. write wrappers for the objective function of those optimizers that can benefit from a smooth penalty returned in the value of the objective function.

If you agree with this solution, I should be able to do these changes in the coming week.

In a second step, I believe that we should replace ```exit_flag``` and ```info``` mechanism with proper exceptions

**Comment by JohannesPfeifer** [comment]
I agree with this solution. You should not worry too much about backward-compatibility with user-defined mode-compute functions as that functionality was broken in 4.4.3.

**Comment by MichelJuillard** [comment]
Thanks @JohannesPfeifer. Do you remember what was broken with user-defined mode-compute. @tholden are you using the user defined ```mode_compute``` when you call  CMA-ES? How did you handle exceptions in that case?

**Comment by JohannesPfeifer** [comment]
We did not correctly account for passing strings to mode_compute. CMA-ES is mode_compute=9. It runs perfectly in the current setup where the base value of the penalty is unchanged

**Comment by tholden** [comment]
I have a custom CMAES which a) takes function handles, b) has more sensible default options, c) also parallelises after the objective function returns nan, (default CMAES handles that case in serial). This is here: https://github.com/tholden/dynare/blob/master/matlab/optimization/cmaes.m
The objective function is wrapped up in a parallel wrapper from here: https://github.com/tholden/dynare/blob/master/matlab/optimization/parallel_wrapper.m which enables the objective to support matrix inputs of params (one possibility per column), and vector outputs of likelihoods, with a parfor loop over the internal calls (and exceptions caught).
Finally, I call via a modified version of dynare_minimize_objective from here: https://github.com/tholden/dynare/blob/master/matlab/optimization/dynare_minimize_objective.m This supports the aforementioned parallel_wrapper, and sets a few more options to sensible values. (Including the bounds which were previously ignored.)

**Comment by JohannesPfeifer** [comment]
Closed by manually merging #1212 @stepan-a Can you please close the ticket

**Comment by stepan-a** [comment]
Done.
